File: worker/src/snapshot.ts
import { aggregate, type AggregateStatus, type BuildDetail, type StatusPayload } from './ci';

export const SLEEP_RUNNING_SECONDS = 120;
export const SLEEP_ATTENTION_SECONDS = 180;
export const SLEEP_FETCH_ERROR_SECONDS = 300;
export const SLEEP_SETTLED_SECONDS = 900;

export const EINK_MAX_WORKFLOWS = 16;

export function sleepSeconds(status: AggregateStatus, isRunning: boolean): number {
  if (isRunning) return SLEEP_RUNNING_SECONDS;
  if (status === 'FAIL' || status === 'UNKNOWN' || status === 'APPROVAL') {
    return SLEEP_ATTENTION_SECONDS;
  }
  if (status === 'CONNECTION_ERROR') return SLEEP_FETCH_ERROR_SECONDS;
  return SLEEP_SETTLED_SECONDS;
}

export interface RepoGlance {
  repo: string;
  status: string;
  workflow_count: number;
  pr_count: number;
  is_running: boolean;
  workflows?: Array<{ workflow: string; status: string }>;
}

export function repoGlances(
  builds: StatusPayload['builds'],
  { includeWorkflows = false }: { includeWorkflows?: boolean } = {},
): RepoGlance[] {
  const byRepo = new Map<string, BuildDetail[]>();
  for (const build of builds) {
    const repo = build.repo || '';
    if (!repo) {
      continue;
    }
    const group = byRepo.get(repo);
    if (group) {
      group.push(build);
    } else {
      byRepo.set(repo, [build]);
    }
  }
  const rows: RepoGlance[] = [];
  for (const [repo, group] of byRepo) {
    const { status, is_running } = aggregate(group);
    let display: string = status;
    if (status === 'NONE' && is_running) {
      const waitingOnly =
        group.some((build) => build.status === 'WAITING') &&
        !group.some((build) => build.status === 'RUNNING');
      display = waitingOnly ? 'WAITING' : 'RUNNING';
    }
    const prRaw = group.find((build) => build.pr_count != null)?.pr_count;
    const prCount = Number(prRaw);
    const row: RepoGlance = {
      repo,
      status: display,
      workflow_count: group.length,
      pr_count: Number.isFinite(prCount) ? prCount : 0,
      is_running,
    };
    if (includeWorkflows) {
      row.workflows = [...group]
        .sort((a, b) =>
          (a.workflow || '').localeCompare(b.workflow || '', undefined, { sensitivity: 'base' }),
        )
        .slice(0, EINK_MAX_WORKFLOWS)
        .map((build) => ({ workflow: build.workflow, status: build.status }));
    }
    rows.push(row);
  }
  rows.sort((a, b) => a.repo.localeCompare(b.repo));
  return rows;
}

export function openPrGlances(
  builds: StatusPayload['builds'],
): Array<{ repo: string; pr_count: number; pr_url: string }> {
  const glances: Array<{ repo: string; pr_count: number; pr_url: string }> = [];
  const seen = new Set<string>();
  for (const build of builds) {
    const repo = build.repo || '';
    if (!repo || seen.has(repo) || build.pr_count == null) {
      continue;
    }
    const numeric = Number(build.pr_count);
    if (!Number.isFinite(numeric) || numeric <= 0) {
      continue;
    }
    seen.add(repo);
    glances.push({
      repo,
      pr_count: numeric,
      pr_url: build.pr_url || `https://github.com/${repo}/pulls`,
    });
  }
  return glances;
}

export function snapshotEtag(payload: StatusPayload): string {
  const body = JSON.stringify({
    builds: payload.builds,
    is_running: payload.is_running,
    open_prs: openPrGlances(payload.builds),
    status: payload.status,
  });
  const digest = hexSha256(body).slice(0, 16);
  return `W/"${digest}"`;
}

export function einkPayload(payload: StatusPayload): {
  type: 'status';
  fetching: false;
  status: StatusPayload['status'];
  is_running: boolean;
  repos: RepoGlance[];
  sleep_seconds: number;
} {
  return {
    type: 'status',
    fetching: false,
    status: payload.status,
    is_running: payload.is_running,
    repos: repoGlances(payload.builds, { includeWorkflows: false }),
    sleep_seconds: sleepSeconds(payload.status, payload.is_running),
  };
}

export function einkRepoPayload(payload: StatusPayload, repo: string) {
  const compact = einkPayload(payload);
  compact.repos = repoGlances(payload.builds, { includeWorkflows: true }).filter(
    (row) => row.repo === repo,
  );
  return compact;
}

export function etagMatches(ifNoneMatch: string | null, etag: string): boolean {
  if (!ifNoneMatch) return false;
  const offered = ifNoneMatch.split(',').map((part) => part.trim()).filter(Boolean);
  return offered.includes('*') || offered.includes(etag);
}

export function statusSnapshotResponse(request: Request, payload: StatusPayload): Response {
  const etag = snapshotEtag(payload);
  const seconds = sleepSeconds(payload.status, payload.is_running);
  const headers = {
    'Cache-Control': 'no-store',
    ETag: etag,
    'Retry-After': String(seconds),
  };
  if (etagMatches(request.headers.get('If-None-Match'), etag)) {
    return new Response(null, { status: 304, headers });
  }

  const url = new URL(request.url);
  const repo = url.searchParams.get('repo');
  const body =
    url.searchParams.get('view') === 'eink'
      ? repo
        ? einkRepoPayload(payload, repo)
        : einkPayload(payload)
      : { ...payload, sleep_seconds: seconds };
  return Response.json(body, { headers });
}

function hexSha256(value: string): string {
  // FNV-1a 64-bit is enough for a weak ETag of this payload; Workers have
  // crypto.subtle but this helper stays sync for unit tests.
  let hash = 0xcbf29ce484222325n;
  for (const byte of new TextEncoder().encode(value)) {
    hash ^= BigInt(byte);
    hash = (hash * 0x100000001b3n) & 0xffffffffffffffffn;
  }
  return hash.toString(16).padStart(16, '0');
}

File: monitor/service/snapshot.py
#!/usr/bin/env python3

"""Snapshot query model for battery and always-on clients.

The hosted Worker (and local WebSocket adapter) expose aggregated CI status
over HTTP. Always-on clients (browser, Mac menu bar) use the full payload.
Battery clients (Xteink X4) sample the same port, then deep-sleep using
``sleep_seconds`` / ``Retry-After``.
"""

from collections.abc import Mapping, Sequence
from hashlib import sha256
from json import dumps
from typing import Any

from monitor.service.aggregator_service import (
    BuildDetail,
    Result,
    builds_in_progress,
    get_status_from_details,
    repo_summaries,
)

# Seconds the device should remain in deep sleep after a successful sample.
# The hub keeps polling on its own cadence; these values are for the radio.
SLEEP_RUNNING_SECONDS = 120
SLEEP_ATTENTION_SECONDS = 180
SLEEP_FETCH_ERROR_SECONDS = 300
SLEEP_SETTLED_SECONDS = 900

EINK_MAX_WORKFLOWS = 16


def sleep_seconds(status: str | Result, *, is_running: bool) -> int:
    """How long a battery client should deep-sleep after this snapshot.

    Running builds win so the panel catches completion. Failures and
    approvals stay shorter than green so the desk still feels live.
    Connection errors back off to avoid a Wi-Fi death spiral.
    """
    value = status.value if isinstance(status, Result) else status
    if is_running:
        return SLEEP_RUNNING_SECONDS
    if value in {Result.FAIL.value, Result.UNKNOWN.value, Result.APPROVAL.value}:
        return SLEEP_ATTENTION_SECONDS
    if value == Result.CONNECTION_ERROR.value:
        return SLEEP_FETCH_ERROR_SECONDS
    return SLEEP_SETTLED_SECONDS


def open_pr_glances(builds: Sequence[Mapping[str, Any]] | None) -> list[dict[str, Any]]:
    glances: list[dict[str, Any]] = []
    seen: set[str] = set()
    for build in builds or []:
        repo = str(build.get("repo") or "")
        count = build.get("pr_count")
        if not repo or repo in seen or count is None:
            continue
        try:
            numeric = int(count)
        except (TypeError, ValueError):
            continue
        if numeric <= 0:
            continue
        seen.add(repo)
        glances.append({
            "repo": repo,
            "pr_count": numeric,
            "pr_url": build.get("pr_url") or f"https://github.com/{repo}/pulls",
        })
    return glances


def snapshot_etag(
    status: str | Result,
    *,
    is_running: bool,
    builds: Sequence[Mapping[str, Any]] | None = None,
) -> str:
    """Weak ETag of glance-relevant fields (not fetch timestamps)."""
    value = status.value if isinstance(status, Result) else status
    body = dumps(
        {
            "builds": list(builds or []),
            "is_running": is_running,
            "open_prs": open_pr_glances(builds),
            "status": value,
        },
        separators=(",", ":"),
        sort_keys=True,
    )
    digest = sha256(body.encode("utf-8")).hexdigest()[:16]
    return f'W/"{digest}"'


def eink_repo_row(summary: Mapping[str, Any], *, include_workflows: bool) -> dict[str, Any]:
    row: dict[str, Any] = {
        "repo": summary["repo"],
        "status": summary["status"],
        "workflow_count": summary["workflow_count"],
        "pr_count": int(summary["pr_count"] or 0),
        "is_running": bool(summary["is_running"]),
    }
    if include_workflows:
        row["workflows"] = [
            {"workflow": item["workflow"], "status": item["status"]}
            for item in list(summary["workflows"])[:EINK_MAX_WORKFLOWS]
        ]
    return row


def eink_repos(builds: Sequence[Mapping[str, Any]] | None, *, include_workflows: bool = False) -> list[dict[str, Any]]:
    details: list[BuildDetail] = []
    for item in builds or []:
        row: BuildDetail = {
            "repo": str(item.get("repo") or ""),
            "workflow": str(item.get("workflow") or ""),
            "status": str(item.get("status") or ""),
            "url": str(item.get("url") or ""),
        }
        pr_count = item.get("pr_count")
        if pr_count is not None:
            row["pr_count"] = int(pr_count)
        pr_url = item.get("pr_url")
        if pr_url:
            row["pr_url"] = str(pr_url)
        details.append(row)
    return [eink_repo_row(summary, include_workflows=include_workflows) for summary in repo_summaries(details)]


def eink_payload(payload: Mapping[str, Any]) -> dict[str, Any]:
    raw_builds = payload.get("builds")
    raw_list = raw_builds if isinstance(raw_builds, list) else []
    details: list[BuildDetail] = []
    for item in raw_list:
        details.append({
            "repo": str(item.get("repo") or ""),
            "workflow": str(item.get("workflow") or ""),
            "status": str(item.get("status") or ""),
            "url": str(item.get("url") or ""),
        })
    rolled = get_status_from_details(details)
    status = rolled.value
    is_running = bool(payload.get("is_running")) or builds_in_progress(details)
    return {
        "type": "status",
        "fetching": False,
        "status": status,
        "is_running": is_running,
        "repos": eink_repos(raw_list, include_workflows=False),
        "sleep_seconds": sleep_seconds(status, is_running=is_running),
    }


def eink_repo_payload(payload: Mapping[str, Any], repo: str) -> dict[str, Any]:
    compact = eink_payload(payload)
    raw_builds = payload.get("builds")
    raw_list = raw_builds if isinstance(raw_builds, list) else []
    match = [row for row in eink_repos(raw_list, include_workflows=True) if row["repo"] == repo]
    compact["repos"] = match
    return compact


def snapshot_headers(
    status: str | Result,
    *,
    is_running: bool,
    builds: Sequence[Mapping[str, Any]] | None = None,
) -> dict[str, str]:
    seconds = sleep_seconds(status, is_running=is_running)
    return {
        "Cache-Control": "no-store",
        "ETag": snapshot_etag(status, is_running=is_running, builds=builds),
        "Retry-After": str(seconds),
    }


def etag_matches(if_none_match: str | None, etag: str) -> bool:
    if not if_none_match:
        return False
    offered = {part.strip() for part in if_none_match.split(",") if part.strip()}
    return "*" in offered or etag in offered
